### ISSUE TITLE
Queuing Notes and/or adding in notes to replays

### ISSUE THREAD

**Issue by Dr-Kakashi** [Zamiell/hanabi-live#442]
I've noticed how notes on cards stay on them when a replay is started.  It's odd seeing f, cm, etc. notes on cards when no clues have been given in a replay.  Would it be possible to have it set so that the notes appear on the cards as the user inputs them when going through the replay?  If on turn 10 I mark f to a card on slot 1, could the note show up on turn 10 in the replay with f in slot 1?

Since we spend so much time reviewing replays, is there a way to set it so that we can add in notes/annotations to the game during the replays?  Like if we're on turn 15 in a replay, it would be nice to be able to mark the start of a finesse for example.

**Comment by Zamiell** [comment]
RE: tracking notes by turn
this has been suggested before and it is problematic.

problem 1 - users may not write notes until later, so the notes may not correspond to the current game state that you are looking at. if you just always show the notes, then it doesn't matter when the user writes it.

problem 2 - if you show every note updated on every turn, then the tooltip could potentially be a giant essay of information that isn't necessarily relevant, especially if the player makes lots of tiny updates to the note for spelling, made a small mistake with the possibilities listed, etc.

**Comment by Zamiell** [comment]
> Since we spend so much time reviewing replays, is there a way to set it so that we can add in notes/annotations to the game during the replays? Like if we're on turn 15 in a replay, it would be nice to be able to mark the start of a finesse for example.

Specify how this is supposed to work

**Comment by Zamiell** [comment]
> Dr_Kakashi🦆🀥Last Thursday at 2:57 AM
> Hi! I actually didn't see that!
> hmm....I was thinking of having a notepad appear alongside the replay
> and you can make annotations as to what plays are being done on what turn
> you could even add notes as to what would ideally be the best play on a specific turn

where does it appear?

**Comment by Dr-Kakashi** [comment]
Sorry for being unable to write here.  I was thinking of maybe have a button that opens another window to write notes.  It would keep the replay the way it is without cluttering it.  Maybe have the button appear above the lobby button that says "note" or "annotations."  That way it's next to the game id as well.

**Comment by Zamiell** [comment]
there's no space above the lobby button - thats where the players hand is to the left of you.

**Comment by Zamiell** [comment]
i could make the lobby button into a home and put the note thing where the terminate button goes though

**Comment by Dr-Kakashi** [comment]
Whichever you feel would be easier to program. That would be a great idea since the terminate button doesn't show in the replay.

**Comment by Zamiell** [comment]
meh not sure how i feel about this feature, i would essentially have to program a psuedo-Google Docs-like collaborative editing into Hanabi Live, and thats a huge work for a questionable benefit

**Comment by zsole2** [comment]
Not for beating on a dead horse, but seeing all notes at the beginning of a replay at once can make really hard to follow the development of the game. Understanding your concerns mentioned above, please consider these comments:

Re problem 1: having notes appear as they were made still would offer an insight on the thought-process of the players.

Re problem 2: leaving the tooltip as is, a log of notes could be kept separately. E.g., in addition to recording them as currently, they could be recorded as a game action, and shown in the game log, e.g. by a different/smaller font. To be useful, though, it would be necessary to have the popup of the game log as a separate scrollable/floating window. 

Anyway, implementing any system for taking account of the timeliness of the notes would immensely help a reviewer to follow better how the game unfolded, making the review an even more useful tool for newcomers to learn the nuances of the conventions (like myself -- although  whenever I actually play Hanabi these days it is in-person and off-line, I'm eager to learn as much as I can by simply watching your expert games)

**Comment by Zamiell** [comment]
one thing you may not realize is that often times players do not write notes until many turns later.
for example, i often write "f" on cards that *are already played*, just so that i won't get confused in future replay dives. or, i write "cm" on a card 5 rounds later after the chop move actually happened because i am watching TV and don't get to looking at the game until it is my turn. etc.

thus, the hypothetical log that you are proposing would likely not be as valuable as you think it would.

**Comment by zsole2** [comment]
I understand this part of your problems. 
But...
In my view, seeing a note later than the event is not worse then seeing it before the event. E.g. I just looked at game no. 48819. By looking at the shuffle and before anything replayed, at least r5 and b4 have comments that seem like being added later getting info on them. Also, m5 has bad info on it., and not corrected later on, as seems occurring sometimes (depends on the specific player, I know). 
If I'll go through the replay, I hopefully will find out how these notes came to those cards, but it will be an effort to deduce it. Sure, it will help me understand the gameplay better, which I'm not against, but definitely will take more time and going back and forth.
I also understand that the great work you put into this site could be spent elsewhere, so I promise not to force this issue further

**Comment by Zamiell** [comment]
since per-turn-notes are being discussed again, interested-parties should re-read this thread, namely:

```
problem 1 - users may not write notes until (much) later, so the notes may not correspond to the current game state that you are looking at. if you just always show the notes, then it doesn't matter when the user writes it.

problem 2 - if you show every note updated on every turn, then the tooltip could potentially be a giant essay of information that isn't necessarily relevant, especially if the player makes lots of tiny updates to the note for spelling, made a small mistake with the possibilities listed, etc.
```

another problem that i did not mention at the time is related to user privacy.

problem 3 - we do not want an immutable record of what a user writes on a card.
users need to have the option to delete their notes for privacy reasons.
otherwise, this will dissuade people from using this feature.
the note is easiest to manage by the user and easiest to conceptualize if it is a single, simple string.
KISS

**Comment by marcmagus** [comment]
Quick suggestion that may satisfy most of the intent of the request while accommodating the noted issues:

1. Automatically prepend the current turn number in [] (or "##:" or whatever) in the editor on note creation. Just go ahead and leave it as part of the plain string which users can edit if they like.

2. Per Discord suggestion, apply a visual indicator on turns which match a turn number format above. Maybe a glow to the note icon rather than a border change, or rotate the note icon until the note is checked like spectators get, so if you're moving quickly in the replay you still see there's an unchecked note.

If people want to keep a record of changes to their note in the note they still can by appending changes (some in my group sometimes do). An exciting but possibly overengineered option would be to allow multiple turn tags in a single note to trigger 2 for people who really want to get into changing thought processes.

**Comment by Zamiell** [comment]
> Automatically prepend the current turn number in [] (or "##:" or whatever) in the editor on note creation. Just go ahead and leave it as part of the plain string which users can edit if they like.

that's actively harmful because by default the turn number won't correspond to the action specified in the note.

**Comment by marcmagus** [comment]
Although it will correspond to the turn in the recap in which the action which triggered the note is visible.

It does require internalizing the idea of "2: f" == "turn 2 we learned this card was finessed turn 1" but that's already a thing, and IME half the time people say the displayed turn number for a clue rather than the turn on which it was actually given.

(Automatically prepend n-1 sounds like it's just asking for trouble)

**Comment by Zamiell** [comment]
i dont think you're following what im saying. the turn on the note wont be equal to n. the turn on the note won't be equal to n - 1. >50% of the time it will be a random number between n and n + m.

**Comment by rfrank27** [comment]
Ignoring the "Automatically prepend the current turn number...." thing. I think having something like "#<turn number>" in a note could change the note icon for that turn in a replay. That a player can mark "[#24] Why didn't you play?", then in a replay the players will have an indication to circle back and talk about that move on turn 24.

I think it's an easy win that doesn't involve any of the timing issues in this thread.

**Comment by Zamiell** [comment]
i would be open to a PR that did that, provided it didn't look too intrusive/bad

**Comment by marcmagus** [comment]
Ah, we are having a disagreement about "by default" because in your experience notes typically get added well after the fact due to multitasking and in my experience >90% of the time they are added immediately (usually turn n+1, sometimes turn n), and when I saw your note I noticed the by-one issue and thought that was your complaint.

In my use case, the rare unknowable lateness might be an acceptable cost for the convenience feature. I see that the more often you have to change the number the less convenient the "convenience" of providing it for you is.
